sharding/peerTypeProvider.go

```diff
@@ -0,0 +1,145 @@
+package sharding
+
+import (
+	"bytes"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/epochStart/notifier"
+)
+
+// TODO: move this component somewhere else and write tests for it
+
+// PeerTypeProvider handles the computation of a peer type
+type PeerTypeProvider struct {
+	nodesCoordinator NodesCoordinator
+	epochHandler     EpochHandler
+	cache            map[string]core.PeerType
+	mutCache         sync.RWMutex
+}
+
+// NewPeerTypeProvider will return a new instance of PeerTypeProvider
+func NewPeerTypeProvider(
+	nodesCoordinator NodesCoordinator,
+	epochHandler EpochHandler,
+	subscriber EpochStartSubscriber,
+) (*PeerTypeProvider, error) {
+	if check.IfNil(nodesCoordinator) {
+		return nil, ErrNilNodesCoordinator
+	}
+	if check.IfNil(epochHandler) {
+		return nil, ErrNilEpochHandler
+	}
+
+	ptp := &PeerTypeProvider{
+		nodesCoordinator: nodesCoordinator,
+		epochHandler:     epochHandler,
+		cache:            make(map[string]core.PeerType),
+		mutCache:         sync.RWMutex{},
+	}
+
+	err := ptp.populateCache(epochHandler.Epoch())
+	if err != nil {
+		return nil, err
+	}
+
+	ptp.registerHandler(subscriber)
```

Comment:
this should have been:
*startOfEpochNotifier*.RegisterHandler(function_to_be_called_when_starting_new_epoch)

and define below the function_to_be_called_when_starting_new_epoch instead of a registerHandler function.

Reply:
done